### ISSUE TITLE
`OffsetDateTime` with `@JsonFormat(without=...)` doesn't seem to work

### ISSUE THREAD

**Issue by ayush-veem** [FasterXML/jackson-datatype-jsr310#98]
Using version 2.8.4, 
OffsetDateTime doesn't seem to serialize to epoch milliseconds correctly with settings that should result in epoch milliseconds.

Given a class with fields like
```
@JsonFormat(shape = JsonFormat.Shape.STRING, pattern = "yyyy-MM-dd'T'HH:mm:ss.SSSXXXX")
private final OffsetDateTime date1;
@JsonFormat(shape = JsonFormat.Shape.NUMBER_INT,
            without = JsonFormat.Feature.WRITE_DATE_TIMESTAMPS_AS_NANOSECONDS)
private final OffsetDateTime date2;
```
and an ObjectMapper configured with
```
WRITE_DATES_AS_TIMESTAMPS(true)
WRITE_DATE_TIMESTAMPS_AS_NANOSECONDS(true)
```
then I'd expect date1 to be serialized in iso format and date2 to be serialized in epoch milliseconds. Instead, date1 is serialized in iso format and date2 is serialized as epoch nanoseconds.

When debugging, I noticed that 
`(provider.isEnabled(SerializationFeature.WRITE_DATE_TIMESTAMPS_AS_NANOSECONDS))` (full context below)
in `InstantSerializerBase.serialize` seems to return true, when I think the JsonFormat annotation on the field should be causing it to return false.
Context
```
if (useTimestamp(provider)) {
            if (provider.isEnabled(SerializationFeature.WRITE_DATE_TIMESTAMPS_AS_NANOSECONDS)) {
                generator.writeNumber(DecimalUtils.toBigDecimal(
                        getEpochSeconds.applyAsLong(value), getNanoseconds.applyAsInt(value)
                ));
                return;
            }
            generator.writeNumber(getEpochMillis.applyAsLong(value));
            return;
        }
        String str;
        
        if (_formatter != null) {
            str = _formatter.format(value);;
        } else if (defaultFormat != null) {
            str = defaultFormat.format(value);;
        } else {
            str = value.toString();
        }
        generator.writeString(str);
```

**Comment by cowtowncoder** [comment]
Please always first upgrade to latest patch of the minor version before reporting. For 2.8, we already have 2.8.10. It may not matter for this particular issue but it is very annoying to sometimes have cases where fix has already been applied.

**Comment by joshuaali** [comment]
With 2.9.4, only the last of the following three tests pass:

```java
import com.fasterxml.jackson.annotation.JsonFormat;
import com.fasterxml.jackson.annotation.JsonProperty;
import com.fasterxml.jackson.databind.ObjectMapper;
import com.fasterxml.jackson.databind.SerializationFeature;
import com.fasterxml.jackson.datatype.jsr310.JavaTimeModule;
import org.junit.Test;

import java.time.Instant;

import static org.hamcrest.Matchers.equalTo;
import static org.junit.Assert.assertThat;

public class JsonFormatTest {
    @Test
    public void testObjectMapper() throws Exception {
        ObjectMapper objectMapper = new ObjectMapper()
                .registerModule(new JavaTimeModule());
        assertThat(objectMapper.writeValueAsString(new Json(Instant.EPOCH)), equalTo("{\"instant\":0}"));
    }

    @Test
    public void testObjectMapperWithWriteDateTimestampsAsNanoseconds() throws Exception {
        ObjectMapper objectMapper = new ObjectMapper()
                .configure(SerializationFeature.WRITE_DATE_TIMESTAMPS_AS_NANOSECONDS, true)
                .registerModule(new JavaTimeModule());
        assertThat(objectMapper.writeValueAsString(new Json(Instant.EPOCH)), equalTo("{\"instant\":0}"));
    }

    @Test
    public void testObjectMapperWithoutWriteDateTimestampsAsNanoseconds() throws Exception {
        ObjectMapper objectMapper = new ObjectMapper()
                .configure(SerializationFeature.WRITE_DATE_TIMESTAMPS_AS_NANOSECONDS, false)
                .registerModule(new JavaTimeModule());
        assertThat(objectMapper.writeValueAsString(new Json(Instant.EPOCH)), equalTo("{\"instant\":0}"));
    }

    static class Json {
        private final Instant instant;

        Json(Instant instant) {
            this.instant = instant;
        }

        @JsonProperty
        @JsonFormat(shape = JsonFormat.Shape.NUMBER_INT, without = JsonFormat.Feature.WRITE_DATE_TIMESTAMPS_AS_NANOSECONDS)
        public Instant getInstant() {
            return instant;
        }
    }
}
```

**Comment by cowtowncoder** [comment]
Thank you for reporting this.

**Comment by joshuaali** [comment]
Had a further look today. `JsonFormat.Feature.WRITE_DATE_TIMESTAMPS_AS_NANOSECONDS` is ignored and always retrieved from the `SerializerProvider` (under `SerializationFeature.WRITE_DATE_TIMESTAMPS_AS_NANOSECONDS`). This is happening everywhere `SerializationFeature.WRITE_DATE_TIMESTAMPS_AS_NANOSECONDS` is used: `DurationSerializer`, `InstantSerializerBase`, `LocalDateTimeSerializer`, `LocalTimeSerializer` and `OffsetTimeSerializer`.

The only feature being honored is `JsonFormat.Feature.WRITE_DATES_WITH_ZONE_ID` and that's only applicable for `ZonedDateTimeSerializer`. This is being read in `JSR310FormattedSerializerBase#createContextual` and passed down via `JSR310FormattedSerializerBase#withFeatures`, which is overriden in `ZonedDateTimeSerializer`, the sole argument being of type `Boolean` (why?).

We can overload this method and add a second `Boolean` argument to represent `JsonFormat.Feature.WRITE_DATE_TIMESTAMPS_AS_NANOSECONDS`. Alternatively, we can treat it similar to `_useTimestamp`, which is a member of the base class, `JSR310FormattedSerializerBase`.

I'm happy to take this on if we have an agreed approach and there's a reasonable timeline (e.g. will this make it into a 2.x release or is it destined for 3.x?)

**Comment by cowtowncoder** [comment]
@joshuaali Since this does not change API, 2.9 is fine (I probably wouldn't bother 2.8 since only micro-patches will be done)... although there may be the question of how to add overloads but keep deprecated constructor for possibility of some users directly extending base classes.

It sounds like your approach wrt how to pass that flag would work and would be happy to help integrate to `2.9` branch.

**Comment by joshuaali** [comment]
Just to clarify, do you prefer overloading `JSR310FormattedSerializerBase#withFeatures` with another `Boolean` parameter and letting each of the 5 subclasses override that or do you prefer keeping all this logic inside `JSR310FormattedSerializerBase` just like what's done with `_useTimestamp`?

**Comment by cowtowncoder** [comment]
@joshuaali I do not remember off-hand what the specific logic is unfortunately, but just from description it sounds like perhaps second approach would be safer. I think this should be rewritten for 3.0 anyway, to simplify, but with 2.9 there is more concern about strict backwards compatibility, including sub-classing.

**Comment by cowtowncoder** [comment]
Fixed in 2.9.5
